DevSkim-DotNet/Microsoft.DevSkim/ScopedRegexOperation.cs

```diff
@@ -0,0 +1,95 @@
+﻿using Microsoft.CST.OAT;
+using Microsoft.CST.OAT.Operations;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.DevSkim
+{
+    public class ScopedRegexOperation : OatOperation
+    {
+        private RegexOperation regexEngine;
+        public ScopedRegexOperation(Analyzer analyzer) : base(Operation.Custom,analyzer)
+        {
+            CustomOperation = "ScopedRegex";
+            regexEngine = new RegexOperation(analyzer);
+            OperationDelegate = ScopedRegexOperationDelegate;
+            ValidationDelegate = ScopedRegexValidationDelegate;
+        }
+
+        public IEnumerable<Violation> ScopedRegexValidationDelegate(CST.OAT.Rule rule, Clause clause)
+        {
+            if (rule is null)
+            {
+                yield return new Violation($"Rule is null", new CST.OAT.Rule("RuleWasNull"));
+                yield break;
+            }
+            if (clause is null)
+            {
+                yield return new Violation($"Rule {rule.Name} has a null clause", rule);
+                yield break;
+            }
+            if (Analyzer is null)
+            {
+                yield return new Violation($"Rule {rule.Name} Clause {clause.Label ?? rule.Clauses.IndexOf(clause).ToString(CultureInfo.InvariantCulture)} ScopedRegexClause requires Analyzer be set.", rule, clause);
+            }
+            if (clause is ScopedRegexClause src)
+            {
+                if (!src.Data?.Any() ?? true)
+                {
+                    yield return new Violation($"Rule {rule.Name} Clause {clause.Label ?? rule.Clauses.IndexOf(clause).ToString(CultureInfo.InvariantCulture)} Must provide some regexes as data.", rule, clause);
+                    yield break;
+                }
+                foreach (var datum in src.Data ?? new List<string>())
+                {
+                    if (regexEngine.StringToRegex(datum, RegexOptions.None) is null)
+                    {
+                        yield return new Violation($"Regex {datum} in Rule {rule.Name} Clause {clause.Label ?? rule.Clauses.IndexOf(clause).ToString(CultureInfo.InvariantCulture)} is not a valid regex.", rule, clause);
+                    }
+                }
+            }
+            else
+            {
+                yield return new Violation($"Rule {rule.Name ?? "Null Rule Name"} clause {clause.Label ?? rule.Clauses.IndexOf(clause).ToString(CultureInfo.InvariantCulture)} is not a ScopedRegexClause", rule, clause);
+            }
+        }
+
+        public OperationResult ScopedRegexOperationDelegate(Clause c, object? state1, object? state2, IEnumerable<ClauseCapture>? captures)
```

Comment:
state2 is never used in this method; Clause could be defined as ScopedRegexClause if no other types are expected to be used for DevSkim at least to avoid an additional type conversion later on line 62; same for state1 -why not define a a TextContainer type to get type checking earlier